apps/frontend/src/app/Solver/GOSolver/polyUB.ts

```diff
@@ -0,0 +1,196 @@
+import { customMapFormula, forEachNodes } from "../../Formula/internal";
+import { OptNode, allOperations } from "../../Formula/optimization";
+import { ConstantNode } from "../../Formula/type";
+import { prod, threshold } from "../../Formula/utils";
+import { assertUnreachable, cartesian } from "../../Util/Util";
+import { ArtifactsBySlot, MinMax, computeFullArtRange, computeNodeRange } from "../common";
+import { Linear } from "./BNBSplitWorker";
+
+export type Polynomial = PolyProd | PolySum | LinTerm
+type LinTerm = { type: 'lin', lin: Linear }
+type PolyProd = { type: 'prod', terms: Polynomial[], $k: number }
+type PolySum = { type: 'sum', terms: Polynomial[], $c: number }
+
+/** Convenience functions modify in-place if possible */
+function constP(n: number): LinTerm { return { type: 'lin', lin: { $c: n } } }
+function readP(k: string): LinTerm { return { type: 'lin', lin: { [k]: 1, $c: 0 } } }
+
+function sumP(...poly: Polynomial[]): Polynomial { return { type: 'sum', terms: poly, $c: 0 } }
+function prodP(...poly: Polynomial[]): Polynomial { return { type: 'prod', terms: poly, $k: 1 } }
+function addP(n: number, poly: Polynomial): Polynomial {
+  switch (poly.type) {
+    case 'lin':
+      poly.lin.$c += n
+      return poly
+    case 'sum':
+      poly.$c += n
+      return poly
+    case 'prod':
+      return sumP(poly, constP(n))
+  }
+}
+function mulP(n: number, poly: Polynomial): Polynomial {
+  switch (poly.type) {
+    case 'lin':
+      for (const [k, v] of Object.entries(poly.lin)) poly.lin[k] = n * v
+      return poly
+    case 'sum':
+      return { type: 'sum', $c: n * poly.$c, terms: poly.terms.map(p => mulP(n, p)) }
+    case 'prod':
+      poly.$k *= n
+      return poly
+  }
+}
+
+function slopePoint(slope: number, x0: number, y0: number, poly: Polynomial): Polynomial {
+  return addP(y0 - slope * x0, mulP(slope, poly))
+}
+function interpolate(x0: number, y0: number, x1: number, y1: number, poly: Polynomial, upper: boolean): Polynomial {
+  if (Math.abs(x0 - x1) < 1e-10)
+    return constP(upper ? Math.max(y0, y1) : Math.min(y0, y1))
+  return slopePoint((y1 - y0) / (x1 - x0), x0, y0, poly)
+}
+
+export function polyUB(nodes: OptNode[], arts: ArtifactsBySlot) {
+  const minMaxes = new Map<OptNode, MinMax>()
+  forEachNodes(nodes, f => {
+    const { operation } = f
+    if (operation === "mul") minMaxes.set(f, { min: NaN, max: NaN })
+    switch (operation) {
```

Comment:
Is this meant to be exhaustive? If so, add `default: assertUnreachable(operation)`